File: leetcode/JavaScript实现/0020.js

// 自己的代码
// 第一想法有一个栈来做
// 遇到左括号将其压入栈，遇到右括号将栈顶弹出，看栈顶是否与右括号匹配
var isValid = function(s) {
    let stack = [], len = s.length;
    for(let i = 0; i < len; ++i) {
        if(s[i] === '(' || s[i] === '[' || s[i] === '{' ) {
            stack.push(s[i]);
        } else if(s[i] === ')') {
            if(stack.pop() !== '(')
                return false;
        } else if(s[i] === ']') {
            if(stack.pop() !== '[')
                return false;
        } else if(s[i] === '}') {
            if(stack.pop() !== '{')
                return false;
        }
    }
    return stack.length === 0 ? true : false;
};

File: leetcode/C++实现/1-99题/0020.cpp
class Solution 
{
public:
    bool isValid(string s) 
    {
      std::stack<char> c;
      for(unsigned int i=0;i<s.length();++i)
      {
          if(s[i]=='('  ||  s[i]=='{' || s[i]=='[' )
          {
              c.push(s[i]);
          }
          else
          {
              if(c.empty())
                  return false;
              char temp=c.top();
               if( (temp=='('&& s[i]==')') ||  ( temp=='[' && s[i]==']' )  || (temp=='{' && s[i]=='}')     )
               {
                   c.pop();
               }
              else
              {
                  return false;
              }
          }
      }
       return c.empty();
    }
};